File: mergeKLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int k = lists.size(),min,flag;
        if(k==0) return NULL;
        ListNode *head = NULL,*p,*q;
        vector<ListNode *>::iterator pos;
        flag = 1;

        while(flag == 1)
        {
            min = INT_MAX;
        	flag = 0;
        	for(vector<ListNode *>::iterator i = lists.begin();i<lists.end();i++)
        	{
        		if(*i == NULL)
        		{
     				continue;
        		}
        		flag = 1;
        		if((*i)->val < min)
        		{
        			min = (*i)->val;
        			pos = i;
        		}
        	}       	
        	if(flag == 1)
        	{
        		*pos = (*pos) -> next;
        		if(head == NULL)
        		{
        			p = new ListNode(min);
        			head = p;
        		}
        		else
        		{
        			q = new ListNode(min);
        			p->next = q;
        			p = q;
        		}
        	}
        }
        return head;
    }
};

File: mergeKLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode mergeKLists(ArrayList<ListNode> lists) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(lists.size() == 0) return null;
        if(lists.size() == 1) return lists.get(0); 
        PriorityQueue heap = new PriorityQueue(lists.size(),new Comparator<ListNode>(){
        	public int compare(ListNode a,ListNode b)
        	{
        		return a.val - b.val;
        	}
        });
        for(ListNode node : lists)
        {
        	if(node != null)
        		heap.add(node);
        }
        ListNode head = new ListNode(0), p = head;
        while(!heap.isEmpty())
        {
        	ListNode node = (ListNode)heap.poll();
        	if(node.next != null)
        		heap.add(node.next);
        	p.next = node;
        	p = node;
        }
        return head.next;
    }
}